cmd/pulumi-resource-docker-buildkit/main.go

```diff
@@ -231,54 +244,78 @@ func (k *dockerBuildkitProvider) dockerBuild(
 		return nil, err
 	}
 
+	var authToken string
 	if !username.IsNull() && !password.IsNull() {
-		cmd := exec.Command(
-			"docker", "login",
-			"-u", username.StringValue(), "--password-stdin",
-			registry["server"].StringValue(),
-		)
-		cmd.Stdin = strings.NewReader(password.StringValue())
-		if err := runCommand(ctx, k.host, urn, cmd); err != nil {
-			return nil, fmt.Errorf("docker login failed: %w", err)
+		auth := types.AuthConfig{
+			Username:      username.StringValue(),
+			Password:      password.StringValue(),
+			ServerAddress: registry["server"].StringValue(),
 		}
+		_, err := k.client.RegistryLogin(ctx, auth)
+		if err != nil {
+			return nil, fmt.Errorf("docker login failed: %v", err)
+		}
+		authConfigBytes, _ := json.Marshal(auth)
+		authToken = base64.URLEncoding.EncodeToString(authConfigBytes)
 	}
 
 	var platforms []string
 	for _, v := range inputs["platforms"].ArrayValue() {
 		platforms = append(platforms, v.StringValue())
 	}
-	cmd := exec.Command(
-		"docker", "buildx", "build",
-		"--platform", strings.Join(platforms, ","),
-		"--cache-from", name,
-		"--cache-to", "type=inline",
-		"-f", filepath.Join(context, dockerfile),
-		"-t", name, "--push",
-		context,
-	)
-	if err := runCommand(ctx, k.host, urn, cmd); err != nil {
-		return nil, fmt.Errorf("docker build failed: %w", err)
+	buildCtx, err := buildContext(context)
+	if err != nil {
+		return nil, fmt.Errorf("could not assemble docker build context for %#v: %v", context, err)
+	}
+	imageBuild := types.ImageBuildOptions{
+		Tags:       []string{name},
+		Dockerfile: dockerfile,
+		Context:    buildCtx,
+		Version:    types.BuilderBuildKit,
+		Remove:     true,
+		CacheFrom:  []string{name},
+		Platform:   strings.Join(platforms, ","),
+		// TODO: I don't know how to get --cache-to=type=internal set here /:
```

Comment:
I think there is a way to set this - the docker CLI does it, after all. Just need to find out what they do.

Reply:
Yeah, but you have to go through `docker buildx build` to get it rather than just `docker build`, which makes me think it might be a thoroughly separate Buildkit API that you've got to go through.

Reply:
Yup yup, I suspect it's like the `--push` argument to `buildx build`, where in the SDK you have to call two client methods instead of setting an option.

Reply:
It's _probably_ as easy as just copying the code from the implementation of `buildx build`? https://github.com/docker/buildx/blob/master/commands/build.go